File: 2a.py
import random
import time
import matplotlib.pyplot as plt

# Generate a sorted array
def generate_sorted_array(n):
    arr = [random.randint(0, 10)]
    for i in range(1, n):
        arr.append(arr[-1] + random.randint(1, 10))
    return arr

# Binary search (with dummy work to amplify timing)
def binary_search(arr, target):
    left, right = 0, len(arr)
    while left < right:
        mid = left + (right - left) // 2
        if arr[mid] == target:
            return True
        if arr[mid] < target:
            # Dummy work to exaggerate runtime difference
            for _ in range(100):
                pass
            left = mid + 1
        else:
            right = mid
    return False

# Benchmark function
def benchmark_binary_search(n,trials=100000):
    arr = generate_sorted_array(n)
    target = arr[-1] + 1  # not present → worst case
    start = time.perf_counter()
    for _ in range(trials):
        binary_search(arr, target)
    end = time.perf_counter()
    total_time = end - start
    avg_time = total_time / trials
    return avg_time

# Run experiments for increasing n
sizes = [100, 500, 1000, 5000, 10000, 20000, 50000]
times = []

for n in sizes:
    avg = benchmark_binary_search(n,trials=20000)  # fewer trials for big n
    print(f"n={n}, avg_time={avg:.2e} s")
    times.append(avg)

# Plot only measured data
plt.figure(figsize=(8,6))
plt.plot(sizes, times, marker='o', color="blue", label="Binary Search (measured)")
plt.xlabel("Array Size (n)")
plt.ylabel("Average Time per Search (seconds)")
plt.title("Binary Search: Average Time vs Array Size")
plt.grid(True)
plt.show()

File: 2a.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

void generateSortedArray(int arr[], int n) {
    arr[0] = rand() % 10;
    for (int i = 1; i < n; i++) {
        arr[i] = arr[i - 1] + (rand() % 10 + 1);
    }
}


int binarySearch(int arr[],int n,int target){
    int left=0,right=n,mid;
    while (left<right)
    {
        mid=left + (right-left)/2;
        if(arr[mid]==target) return 1;
        if(arr[mid]<target){
            for (volatile int j=0; j<100; j++);
            left=mid+1;
        }else{
            right=mid;
        }
    }
    return 0;
}

int main(){
    int n;
    printf("Enter size of array\n");
    scanf("%d",&n);
    
    int *arr = (int *)malloc(n * sizeof(int));
    if (arr == NULL) {
        printf("Memory allocation failed\n");
        return 1;
    }
    
    generateSortedArray(arr, n);
    clock_t start =clock();
    
    for(int i=0;i<1000000;i++){
        binarySearch(arr,n,arr[n-1]+1);
    }
    clock_t end =clock();

    double time_taken = (double)(end -start) /CLOCKS_PER_SEC;
    double avg_time = time_taken/100000.0;
    printf("Total time: %f s,Average per search: %f s\n", time_taken,avg_time);

    return 0;
}